File: daiko/services/geo.py
"""位置情報ヘルパー — 距離計算と近隣ドライバー抽出."""
from __future__ import annotations

import math


def haversine_km(lat1: float, lng1: float, lat2: float, lng2: float) -> float:
    """2点間の概算距離（km）."""
    r = 6371.0
    p1, p2 = math.radians(lat1), math.radians(lat2)
    dp = math.radians(lat2 - lat1)
    dl = math.radians(lng2 - lng1)
    a = math.sin(dp / 2) ** 2 + math.cos(p1) * math.cos(p2) * math.sin(dl / 2) ** 2
    return 2 * r * math.asin(math.sqrt(a))

File: yorisoi/src/lib/geo.ts
/**
 * 位置計算ユーティリティ（プラットフォーム非依存）。
 */
import type { GeoPoint } from "@/types/location";

/** 2 点間の距離（メートル）をハーバサイン公式で求める */
export function distanceMeters(a: GeoPoint, b: GeoPoint): number {
  const R = 6_371_000; // 地球半径(m)
  const toRad = (deg: number) => (deg * Math.PI) / 180;
  const dLat = toRad(b.lat - a.lat);
  const dLng = toRad(b.lng - a.lng);
  const lat1 = toRad(a.lat);
  const lat2 = toRad(b.lat);
  const h =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLng / 2) ** 2;
  return 2 * R * Math.asin(Math.sqrt(h));
}

/** 点 p が中心 center・半径 radius(m) の円内にあるか */
export function isWithin(
  p: GeoPoint,
  center: GeoPoint,
  radiusMeters: number,
): boolean {
  return distanceMeters(p, center) <= radiusMeters;
}
